cassovary-core/src/main/scala/com/twitter/cassovary/graph/MemoryMappedDirectedGraph.scala

```diff
@@ -1,34 +1,52 @@
-import it.unimi.dsi.fastutil.ints.{IntArrayList, IntList}
+import com.twitter.cassovary.util.FastUtilUtils
+import com.twitter.cassovary.util.io.MemoryMappedIntLongSource
+import it.unimi.dsi.fastutil.ints.IntArrayList
 
+import scala.collection.mutable.ArrayBuffer
 import scala.io.Source
 
 /**
- * A graph which reads edge data from a memory mapped file.  There is no object overhead per node: the memory
- * used for n nodes and m edges with both in-neighbor and out-neighbor access is exactly 8 + 16*n + 8*m bytes.
- * Also, Loading is very fast because no parsing of text is required.  Loading time is exactly the time it takes the operating system
- * to map data from disk into memory.
- *
- * Nodes are numbered sequentially from 0 to nodeCount - 1 and must be a range of this form (i.e. nodeCount == maxNodeId + 1).
- * When transforming a graph where nodeCount <= maxNodeId to this format, new nodes with no neighbors will be implicitly created.
+ * A graph which reads edge data from a memory mapped file.  There is no object overhead per
+ * node: the memory used for n nodes and m edges with both in-neighbor and out-neighbor access is
+ * exactly 16 + 16*n + 8*m bytes. Also, loading is very fast because no parsing of text is required.
+ * Loading time is exactly the time it takes the operating system to map data from disk into
+ * memory.  Nodes are numbered sequentially from 0 to nodeCount - 1 and must be a range of this
+ * form (i.e. nodeCount == maxNodeId + 1).
  *
- * Currently only supports storing both in-neighbors and out-neighbors of nodes (StoredGraphDir.BothInOut).
- * The binary format is currently subject to change.
+ * When transforming a graph where nodeCount <= maxNodeId
+ * to this format, new nodes with no neighbors will be implicitly created.  Currently only supports
+ * storing both in-neighbors and out-neighbors of nodes (StoredGraphDir.BothInOut). The binary
+ * format is currently subject to change.  Node objects are created on demand when getNodeById is
+ * called.
  */
 
 /* Storage format
 byteCount  data
-4          (reserved, later use for versioning or indicating undirected vs directed)
-4          n (i. e. the number of nodes)
-8*(n+1)    Offsets into out-neighbor data. Index i (a Long) points to the out-neighbor data of node i.
+8          (reserved, later use for versioning or indicating undirected vs directed)
+8          n (i. e. the number of nodes).  Currently must be less than 2^31.
+8*(n+1)    Offsets into out-neighbor data. Index i (a Long) points to the out-neighbor data of
+           node i.  The out-neighbor data must be stored in sequential order by id, as the
+           outegree of node i is computed from the difference in offset between node i+1 and node i.
            Index n is needed to compute the outdegree of node n - 1.
 8*(n+1)    Offsets into in-neighbor data (Longs) (Same interpretation as out-neighbor offsets)
 m          out-neighbor data
```

Comment:
Wow, good catch.  I don't know how I missed that before.
(And I think it should be 4*m, because we're using 4 byte node ids.)